fathom/src/core/binary.rs

```diff
@@ -1,17 +1,188 @@
 //! Binary semantics of the data description language
 
 use std::collections::HashMap;
-use std::io::{self, Read, Seek, SeekFrom};
+use std::convert::TryFrom;
+use std::slice::SliceIndex;
 use std::sync::Arc;
 
 use crate::core::semantics::{self, ArcValue, Elim, Head, Value};
 use crate::core::{Const, Prim, UIntStyle};
 use crate::env::{EnvLen, SliceEnv};
 
+#[derive(Clone, Debug)]
+pub enum ReadError {
+    InvalidFormat,
+    InvalidValue,
+    UnwrappedNone,
+    ReadFailFormat,
+    CondFailure,
+    SeekOutOfBounds,
+    UnexpectedEndOfBuffer,
+    PositionOverflow,
+}
+
+/// A buffer that starts at an offset into a larger buffer.
+///
+/// ```text
+///  ┌───────────────────┬─────────────────────────────────────────────────────────────────────┐
+///  │                   │                                data                                 │
+///  │                   ├─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬ ─ ─ ─ ─ ─ ─ ─ ┬─────┤
+///  │                   │  ?  │  ?  │  ?  │  ?  │  ?  │  ?  │  ?  │  ?  │               │  ?  |
+///  └───────────────────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴ ─ ─ ─ ─ ─ ─ ─ ┴─────┘
+///  0       ...         n    n+1   n+2   n+3   n+4   n+5   n+6   n+7   n+8     ...    n+m-1  n+m
+/// -n       ...         0     1     2     3     4     5     6     7     8      ...      m-1   m
+///  │                   │                                                                     │
+///  └─ Buffer::offset ─>└─────────────────────── Buffer::remaining_len ──────────────────────>│
+///  │                   │                                   │                                 │
+///  │                   └── BufferReader::relative_offset ─>│                                 │
+///  │                                                       │                                 │
+///  └─────────────── BufferReader::offset ─────────────────>└── BufferReader::remaining_len ─>│
+/// ```
+#[derive(Copy, Clone)]
+pub struct Buffer<'data> {
+    /// Offset from the starting position.
+    offset: usize,
```

Comment:
Minor: `offset` kinda feels mildly confusing because at first glance it looks like it would be an offset within `data`. Perhaps `start_pos`, `base_pos`, `base_offset` would be clearer?

Reply:
Do you think it would be weird to change the name of `BufferReader::offset` to something similar?

Reply:
Or maybe we could call it `BufferReader::absolute_offset` or something 🤔

Reply:
I don't think it would be weird.